---
title: "metagenomic_workflow"
author: "Daphne Garcia"
date: "2025-05-27"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

## Goals

The goal of this script is to assemble and process the metagenomic data
from floral, crop, and pollen provisions collected by Vivianna Sanchez
for her dissertation. I will be using some code from Sophia's rotation,
found in
`/local/workdir/sna49/Hendry_Rotation_SA/Acinetobacter/Acinetobacter_Lab_`
`Notebook/Acinetobacter_Lab_Notebook.Rmd`

## Steps

Copied from Katie's workflow outlined in her paper:

1.  FastQC quality analysis
2.  Trimmomatic adapter trimming
3.  Bowtie2 read filtering
4.  MetaSpades assembly
5.  CheckM host contamination
6.  Phyloflash taxonomy composition

The inputs and outputs will be as follows:

+----------+----------+----------+----------+----------+----------+
| Program  | Function | Input    | Output   | Input    | Output   |
|          |          |          |          | path     | path     |
+==========+==========+==========+==========+==========+==========+
| FastQC   | Quality  | raw      | .html    | ./0      | ./01     |
|          | control  | reads    | analysis | 1_       | \_fas    |
|          |          |          |          | RawReads | tQC_resu |
|          |          |          |          |          | lts_     |
|          |          |          |          |          | pre_trim |
+----------+----------+----------+----------+----------+----------+
| Tr       | adapter  | raw      | pair     | ./0      | ./02\_   |
| i        | trimming | reads    | ed/      | 1_       | trim     |
| mmomatic |          |          | unpaired | RawReads | momatic/ |
|          |          |          | adap     |          |          |
|          |          |          | ter      |          |          |
|          |          |          | -trimmed |          |          |
|          |          |          | reads    |          |          |
+----------+----------+----------+----------+----------+----------+
| Bowtie2  | align to | trimmed  | u        | ./02\_   | ./03     |
|          | r        | and      | naligned | tr       | _Bowtie2 |
|          | eference | grouped  | reads as | im_unzip | \_hos    |
|          |          | fastq    | fastq    | pe       | t_filter |
|          |          | reads    |          | d_and_ho |          |
|          |          |          |          | st       |          |
|          |          |          |          | -grouped |          |
|          |          |          |          |          |          |
|          |          |          |          | ./03     |          |
|          |          |          |          | \_F      |          |
|          |          |          |          | lower_ho |          |
|          |          |          |          | st       |          |
|          |          |          |          | _genomes |          |
+----------+----------+----------+----------+----------+----------+
| M        | Assemble | host     |          | /03      |          |
| e        | reads    | -removed |          | \_B      |          |
| taSpades |          | R1 and   |          | owtie2_h |          |
|          |          | R2 files |          | os       |          |
|          |          |          |          | t_filter |          |
+----------+----------+----------+----------+----------+----------+
| Quast    | check    | (on      |  NA      |   NA     |    NA    |
|          | assembly | vivi's   |          |          |          |
|          | quality  | jupyter) |          |          |          |
+----------+----------+----------+----------+----------+----------+
| MetaBat2 |  contig  |          |          |          |          |
|          | binning  |          |          |          |          |
|          |          |          |          |          |          |
+----------+----------+----------+----------+----------+----------+
| CheckM   | host     |          |          |          |          |
|          | conta    |          |          |          |          |
|          | mination |          |          |          |          |
+----------+----------+----------+----------+----------+----------+
| p        | t        | (vivi    |          |          |          |
| h        | axonomic | used M   |          |          |          |
| yloflash | com      | etaPhlAn |          |          |          |
| pipeline | position |          |          |          |          |
+----------+----------+----------+----------+----------+----------+

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Acquiring Data

May 27, 2025 I copied Vivi's metagenomics raw reads from her folder to
mine:
`cp /local/workdir/vas75/metaspades_flowers/metagenomic /local/workdir/Daphne/`
`Floral-metagenomics/data/01_raw_reads`

This took \~ 30 seconds

I noticed a pattern with the filenames: all start with
13702_32540_1797XX The X's ranging from **38** (R1 and R2) to **93** (R1
and R2).


### 1. FastQC Quality analysis

In order to assess the quality of our reads, I ran fastQC on all
metagenomic samples, and multiQC on the total

FastQC:

```{bash first_fastqc, eval = FALSE}
## NOTE: Since this chunk is in bash, run with ctrl + alt + enter

# download fastqc from biohpc
export PATH=/programs/FastQC-0.12.1:$PATH

#First, write a list of all of the metagenomic filenames, removing "fastq.gz"
cd data/01_raw_reads/metagenomic
ls *.fastq.gz | sed 's/\.fastq\.gz$//' > samples.txt

# Loop over the list of filenames and perform fastQC, outputting to data/01_fastq...
  for sample in $(cat ./samples.txt)
do
        fastqc $(echo ${sample}.fastq.gz) -o ../../01_fastQC_results_pre_trim
done

```

This step took 42 min

MultiQC:

```{bash first_multiQC, eval = FALSE}
## NOTE: Since this chunk is in bash, run with ctrl + alt + enter
    
# download multiqc from biohpc
export PYTHONPATH=/programs/multiqc-1.15/lib64/python3.9/site-packages:/programs/multiqc-1.15/lib/python3.9/site-packages
export PATH=/programs/multiqc-1.15/bin:$PATH
 
# Run multiqc on all of the outputs from fastqc  
cd /local/workdir/Daphne/Floral_metagenomics/data/01_fastQC_results_pre_trim
multiqc *_fastqc.zip

```

The fastQC and multiQC outputs are in
`/local/workdir/Daphne/Floral_metagenomics/data/01_fastQC_results_pre_trim`,
and both matched with Sophia's outputs.


### 2. Trimmomatic adapter trim

After the initial quality control, I ran trimmomatic to remove the
Nextera adapters detected by the fastQC

```{bash, trimmomatic, eval = FALSE}
## NOTE: Since this chunk is in bash, run with ctrl + alt + enter

RANDOM=42

#set workdir to metagenomic to access the fastq.gz files
cd /local/workdir/Daphne/Floral_metagenomics/data/01_raw_reads/metagenomic

# init a variable the path to Trimmomatic JAR file, copied from BioHPC website
trimmomatic_jar="/programs/trimmomatic/trimmomatic-0.39.jar" 

#important to note what primers do we have.. according to multiqc we have the nextera primers 
    for f2 in *_R2.fastq.gz; do
    echo "THIS IS $f2" 
    printf "\n"
    f1="${f2%%_R2.fastq.gz}_R1.fastq.gz"
    output_paired_1="paired_${f1}"
    output_unpaired_1="unpaired_output_${f1}"
    output_paired_2="paired_${f2}"
    output_unpaired_2="unpaired_output_${f2}"
    java -jar "$trimmomatic_jar" PE -phred33 "$f1" "$f2" \
        "$output_paired_1" "$output_unpaired_1" "$output_paired_2" "$output_unpaired_2" \
        ILLUMINACLIP:/programs/trimmomatic/adapters/NexteraPE-PE.fa:2:30:10 \
        LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36
done

#This took 5 hours
#The output will be in /01_raw_reads, but the following lines will move it to 02_trimmomatic

# move trimmomatic results to Floral_metagenomics/data/02_trimmomatic
mv unpaired* ../../02_trimmomatic/
mv paired* ../../02_trimmomatic/

```

Outputted paired and unpaired files for R1 and R2 of 38-92 are in
`/local/workdir/Daphne/Floral_metagenomics/data/02_trimmomatic/`

#### 2.1 FastQC Quality analysis post-trim

After the trimmomatic removed the adapters, let's run the quality
analysis again on the paired reads to see if the quality improved

```{bash post-trim_fastqc_and_multiqc, eval = FALSE}

## NOTE: Since this chunk is in bash, run with ctrl + alt + enter

# download fastqc from biohpc
export PATH=/programs/FastQC-0.12.1:$PATH

#Go to directory with trimmomatic paired and unpaired
cd /local/workdir/Daphne/Floral_metagenomics/data/02_trimmomatic
 
# loop through paired files to do fastqc to see if our trimming is better 
for sample in paired_*.fastq.gz
do
        fastqc $(echo ${sample}) -o ../02_fastQC_results_post_trim
done

# This took 23 minutes

#go to directory with the paired trimmed fastqc results
cd ../02_fastQC_results_post_trim

# Run multiqc on all of the outputs from fastqc
export PYTHONPATH=/programs/multiqc-1.15/lib64/python3.9/site-packages:/programs/multiqc-1.15/lib/python3.9/site-packages
export PATH=/programs/multiqc-1.15/bin:$PATH

multiqc *_fastqc.zip

#i think the quality scores look good now? after consulting with the internet i think its fine?
```

outputted fastQC files for every sample and multiQC for total are in
`/local/workdir/Daphne/Floral_metagenomics/data/02_fastQC_results_post_trim`

#### 2.2 Grouping trimmed reads by host genome

Now I need to unzip all of the samples and group them the host they were
sampled on (ex: dandelion, apple, trefoil). This way, we can filter out
their host genome with bowtie2 using the corresponding reference genome.
Their metadata is in
`workdir/Daphne/Floral_metagenomics/data/metadata_host_flower_genomes.xlsx`

following the metadata, matched the number 38-93 to the host species
using it's 'FN' tag in the filenames.

```{bash group_by_host, eval = FALSE}
cd /local/workdir/Daphne/Floral_metagenomics/data

# copy all paired reads from 02_trimmomatic directory into the new one
cp 02_trimmomatic/paired* 02_trim_unzipped_and_host-grouped/

cd 02_trim_unzipped_and_host-grouped/

gunzip paired*

## From here, I manually moved each file into a corresponding subdirectory matching it to the right reference host chosen for it

## This is based on '22FN0XX' in the filenames matched with metadata  Floral_metagenomics/data/metadata_sequencing.xlsx

```

Output is all of the trimmed paired unzipped fastq's in corresponding
subdirectories according to their host flower in
`Floral_metagenomics/data/02_trim_unzipped_and_host-grouped`


### 3. Bowtie2

We're gonna run the trimmomatic paired and unpaired trimmed reads
through Bowtie2 to align the reads to their corresponding host genome.
The samples were washed off of flower petals, (info in
`Floral_metagenomics/data/metadata_host_flower_genomes.xlsx`) so for
each sample I will be aligning to corresponding genomes from
spreadsheet:
/local/workdir/Daphne/Floral_metagenomics/data/metadata_host_flower_genomes.xlsx

#### 3.1 Download/build reference indices

First, I need to create reference indices for every host genome that
I'll be using. I found the RefSeq's online through NCBI genome finder,
downloaded the FTP `_genomic.fna.gz` file to my computer, then
transfered it to the server.

| Number | reference genome     |
|--------|----------------------|
| 1      | Cichorium intybus    |
| 2      | Malus domestica      |
| 3      | Lotus japonicus      |
| 4      | Ranunculus sardous   |
| 5      | Trifolium repens     |
| 6      | Matricaria discoidea |
| 7      | Alliaria petiolata   |
| 8      | Solidago caesia      |
| 9      | Lonicera japonica    |
| 10     | Linaria vulgaris     |

```{bash bowtie2_build_reference_indices, eval=FALSE}

cd /local/workdir/Daphne/Floral_metagenomics/data/03_Flower_host_genomes

export PATH=/programs/bowtie2-2.5.1-linux-x86_64:$PATH

#Index for Cichorium intybus (dandelion, chickory, hawkweed)
bowtie2-build GCA_023525715.1_ASM2352571v1_genomic.fna.gz Chichorium_indices/Chichorium_index

#RefSeq of Malus domestica (apple)
bowtie2-build GCF_042453785.1_GDT2T_hap1_genomic.fna.gz Malus_indices/Malus_index

#RefSeq of Lotus japonicus (trefoil)
bowtie2-build GCF_012489685.1_LjGifu_v1.2_genomic.fna.gz Lotus_indices/Lotus_index

#RefSeq of Ranunculus sardous (buttercup)
bowtie2-build GCA_965125545.1_dmRanSard1.hap1.1_genomic.fna.gz Ranculus_indices/Ranunculus_index

#RefSeq of Trifolium repens (clover)
bowtie2-build GCA_030408175.1_UTM_Trep_v1.0_genomic.fna.gz Trifolium_indices/Trifolium_index

#RefSeq of Matricaria discoidea (yarrow + ox eye daisy)
bowtie2-build GCA_964276665.1_daMatDisc1.hap1.1_genomic.fna.gz Matricaria_indices/Matricaria_index

#Refseq of Alliaria petiolata (garlic mustard)
bowtie2-build GCA_049307985.1_ddAllPeti2_p1.1_genomic.fna.gz Alliaria_indices/Alliaria_index

#Refseq of Solidago caesia (golden rod + aster)
bowtie2-build GCA_028566075.1_ASM2856607v1_genomic.fna.gz Solidago_indices/Solidago_index
#Refseq of Lonicera japonica (honeysuckle)
bowtie2-build GCA_021464415.1_ASM2146441v1_genomic.fna.gz Lonicera_indices/Lonicera_index

#Refseq of Linaria vulgaris (toadflax)
bowtie2-build GCA_948329865.1_daLinVulg1.1_genomic.fna.gz Linaria_indices/Linaria_index

#inspect content of each database (replace Chichorium with other names)
bowtie2-inspect -n Chichorium_indices/Chichorium_index
  #Usually its a long list of 100+ scaffolds and genomes, idk what it means though

```

Output is is in 10 directories under
`Floral_metagenomics/data/03_Flower_host_genomes/` (1-10), each with
index files composing the index for that reference genome

#### 3.2 Bowtie host filtering

Using this Bowtie loop, we can loop through all of the samples in a
genome's subdirectory and compare how many of the reads align to the
reference. We input the samples saved in
`02_trim_unzipped_and_host-grouped`, the reference genome in
`03_Flower_host_genomes` and output into `03_Bowtie2_host_filter/`.

I need to modify the three variables: `index`, `samples_folder`, and
`output_folder` in the first lines to change which host genome's samples
we are using

```{bash bowtie2_loop, eval = FALSE}

cd /local/workdir/Daphne/Floral_metagenomics/data/
export PATH=/programs/bowtie2-2.5.1-linux-x86_64:$PATH

## MODIFY THESE NAMES
index="03_Flower_host_genomes/Chichorium_indices/Chichorium_index"
samples_folder="02_trim_unzipped_and_host-grouped/1_Chichorium_samples/"
output_folder="03_Bowtie2_host_filter/1_Chichorium/"
 
for i in "$samples_folder"*R1.fastq; do  
  base=$(basename "$i" _R1.fastq)  # just filename without path
  output_dir="${output_folder}${base}"; # each sample's path inside bowtie dir
  mkdir $output_dir
  bowtie2 --sensitive-local -p 32 --seed 4 -x $index -1 $i -2 ${i/%R1.fastq/R2.fastq} -S "$output_dir/SAMPLE_mapped_and_unmapped.sam"; 
done 

# --sensitive-local (not very-sensitive) because our ref host genomes are usually different genuses of the actual host
# -p is the number of parallel threads 
#--seed 4 is a random number to make it so that work is reproducible 
# -x The basename of the index for the reference genome variable $index
# -1 and -2 are the paired inputs R1 and R2
# -S Summary of index settings and input names/lengths

```

Outputs are `paired_1370.../SAMPLE_mapped_and_unmapped.sam` file within
`03_Bowtie2_host_filter/#_host-genome/`

The percent alignment for each sample (38-93) are in the fourth sheet of
`workdir/Daphne/Floral_metagenomics/data/metadata_host_flower_genomes.xlsx`

#### 3.3 SamTools and bedtools modifications

The unmapped reads must be separated, since these are the non-flower reads, and 
Sophia and Katie both had these lines to prepare for the SpAdes alignment.
I copied Sophia's code and created a for loop for every sample in the 
host-genome's subdirectory.

```{bash sam_bed_after_bowtie2, eval = FALSE}

cd /local/workdir/Daphne/Floral_metagenomics/data/

for sample in 03_Bowtie2_host_filter/8_Solidago/*/; do
  cd $sample
  pwd

  #Convert file .sam to .bam
  touch SAMPLE_mapped_and_unmapped.bam
  samtools view -bS SAMPLE_mapped_and_unmapped.sam > SAMPLE_mapped_and_unmapped.bam
  
  #filter required unmapped reads get unmapped pairs (both ends unmapped)
  samtools view -b -f 12 -F 256 SAMPLE_mapped_and_unmapped.bam > SAMPLE_bothReadsUnmapped.bam 
  #-f 12     Extract only (-f) alignments with both reads unmapped
  #-F 256   Do not(-F) extract alignments which are: <not primary alignment>
  
  # split paired-end reads into separated fastq files .._r1 .._r2
  samtools sort -n -m 5G -@ 2 SAMPLE_bothReadsUnmapped.bam -o SAMPLE_bothReadsUnmapped_sorted.bam
  
  # sort bam file by read name (-n) to have paired reads next to each other as required by bedtools
  samtools fastq -@ 8 SAMPLE_bothReadsUnmapped_sorted.bam -1 SAMPLE_host_removed_R1.fastq.gz -2 SAMPLE_host_removed_R2.fastq.gz -0 /dev/null -s /dev/null -n
  
  #Convert bam to fasstq
  bedtools bamtofastq -i SAMPLE_bothReadsUnmapped_sorted.bam -fq SAMPLE_host_removed_R1.fastq -fq2 SAMPLE_host_removed_R2.fastq
  
  cd - > /dev/null # return to previous dir
done

```

This outputs 6 new files into
`Floral_metagenomics/data/03_Bowtie2_host_filter/`
`(# host name)/paired...(sample name)/` - both reads mapped and unmapped
(sam -\> bam -\> sorted) - host removed R1, R2 (fastq.gz -\> .fastq)

the host-removed R1 and R2 files will be subsequently used in the
metaspades assembly


### 4. MetaSpAdes assembly

Using the Unmapped reads output from the previous chunk, we will run
each sample through metaspades and assemble all of the genomes.

This script is based off of Sophia's second metaspades, after bowtie
Note: My parameter choices on thread and spades --meta instead of metaspades 
are in my lab notebook on benchling
Note: I created directory 04_metaspades and all 10 subdirectories by hand

```{bash Spades_loop, eval=FALSE}

export PATH=/programs/SPAdes-4.0.0/bin:$PATH
cd /local/workdir/Daphne/Floral_metagenomics/data

### MODIFY THIS VARIABLE
host_gen="8_Solidago"
for sample in 03_Bowtie2_host_filter/"$host_gen"/*/; do
  cd $sample
  pwd
  base=$(basename "$sample")
  mkdir "../../../04_metaspades/"$host_gen"/"$base
  spades.py -1 SAMPLE_host_removed_R1.fastq -2 SAMPLE_host_removed_R2.fastq --meta -t 32 -m 200 -o "../../../04_metaspades/"$host_gen"/"$base
  cd - > /dev/null  # return to previous dir
  pwd 
done 

#in this command  
  # -1 = file with forward reads
  # -2 = file with reverse reads
  # -t = threads sets the number of processors to use we are using 16 bc thats the default
  # -m = memory limit in Gb. SPAdes terminates this if it reaches this limit so we are setting it to 200Gb to be safe but the default is 250 Gb
  # -o = output directory to use the default is in the current directory
```

Outputs according to spades manual https://currentprotocols.onlinelibrary.wiley.com/doi/full/10.1002/cpbi.102:
1. contigs.fasta — resulting contig sequences in FASTA format;
2. scaffolds.fasta — resulting scaffold sequences in FASTA format;
3. assembly_graph.gfa — assembly graph and scaffolds paths in GFA 1.0 format;
4. assembly_graph.fastg — assembly graph in FASTG format;
5. contigs.paths — paths in the assembly graph corresponding to contigs.fasta;
6. scaffolds.paths — paths in the assembly graph corresponding to scaffolds.fasta;
7. spades.log — file with all log messages.


### 5. MetaQuast check

This isn't a step in the process, but is important to start checking if acinetobacter is actually present in these samples. Metaquast is like FastQC but for contigs, it "evaluates and compares metagenome assemblies based on alignments to close references".

Therefore, we provide acinetobacter reference genomes and it more accurately bins those contigs into MAGS, and tells us how many contigs map to acinetobacter. It also provides stats on sample quality

I built the references with 12 strains from Sophia's folder `/local/workdir/sna49/Hendry_Rotation_SA/Acinetobacter/Data/metaspades_floral/metagenomic/original_reads/pipeline/01_trimmomatic_paired/references` plus 13 more from 
Vivi's RAST

TO RUN: modify the variable host_name for which of the 10 flower hosts you will 
use
```{bash MetaQuast, eval = FALSE}

export PYTHONPATH=/programs/quast-5.2.0/lib64/python3.9/site-packages:/programs/quast-5.2.0/lib/python3.9/site-packages
export PATH=/programs/quast-5.2.0/bin:$PATH

#This list contains paths to all acineto reference genomes, from the 04_metaspades sample's directory
list="../../../05_Quast_reference_acineto_genomes/A39.fna,../../../05_Quast_reference_acineto_genomes/ANC4215.fna,../../../05_Quast_reference_acineto_genomes/ANC4422.fna,../../../05_Quast_reference_acineto_genomes/B10A.fna,../../../05_Quast_reference_acineto_genomes/B5B.fna,../../../05_Quast_reference_acineto_genomes/BB226.fna,../../../05_Quast_reference_acineto_genomes/BB362.fna,../../../05_Quast_reference_acineto_genomes/BRTC1.fna,../../../05_Quast_reference_acineto_genomes/CIP110305.fna,../../../05_Quast_reference_acineto_genomes/CIP110357.fna,../../../05_Quast_reference_acineto_genomes/CIP110549.fna,../../../05_Quast_reference_acineto_genomes/DSM14961.fna,../../../05_Quast_reference_acineto_genomes/DSM14964.fna,../../../05_Quast_reference_acineto_genomes/EC031.fna,../../../05_Quast_reference_acineto_genomes/EC034.fna,../../../05_Quast_reference_acineto_genomes/EC115.fna,../../../05_Quast_reference_acineto_genomes/EC24.fna,../../../05_Quast_reference_acineto_genomes/FNA11.fna,../../../05_Quast_reference_acineto_genomes/FNA3.fna,../../../05_Quast_reference_acineto_genomes/NCTC5866.fna,../../../05_Quast_reference_acineto_genomes/NIPH706.fna,../../../05_Quast_reference_acineto_genomes/NIPH991.fna,../../../05_Quast_reference_acineto_genomes/SCC474reference.fna,../../../05_Quast_reference_acineto_genomes/SCC477.fna,../../../05_Quast_reference_acineto_genomes/SM1.fna"

# MODIFY this host_name for the samples you're running
host_name="9_Lonicera"
for sample in 04_metaspades"/$host_name/"*; do
  #loop through samples in a flower-host group in metaspades folder
  cd $sample
  name="$(basename "$sample" | cut -d'_' -f1-4)" 
  echo $name
  # run metaquast, output to 05_metaQUAST, and use $list variable as reference
  metaquast.py -o ../../../05_metaQUAST/"$host_name/""$name"_metaQuast contigs.fasta -r "$list" 
  cd - > /dev/null  # return to previous dir
  pwd 
done 
```
The output is 
`Floral_metagenomics/data/05_metaquast/"host_group"/"sample"/*`


Next steps: Since there is so little acinetobacter, I can either focus on one type of host that has acinetobacter, or focus on the whole community's 

#### 5.1 Metaquast without Acinetobacter references
I've decided to Run MetaQUAST again, but without using any references. I'll Use a database, like SILVA instead.

To run this faster, I am splitting the 10 groups into 3 larger arrays, and tmux automating this process in 3 different groups

```{bash metaQuast_no_reference, eval = FALSE}

declare -a array1=(
[0]=1_Chichorium
)
declare -a array2=(
[0]=2_Malus
[1]=4_Ranunculus
[2]=7_Alliaria
[3]=10_Linaria
)
declare -a array3=(
[0]=3_Lotus
[1]=5_Trifolium
[2]=8_Solidago
[3]=6_Matricaria
[4]=9_Lonicera
)
export PYTHONPATH=/programs/quast-5.2.0/lib64/python3.9/site-packages:/programs/quast-5.2.0/lib/python3.9/site-packages
export PATH=/programs/quast-5.2.0/bin:$PATH

for host_name in ${array3[@]}; do
  echo $host_name
  for sample in 04_metaspades"/$host_name/"*; do
    cd $sample
    pwd
    name="$(basename "$sample" | cut -d'_' -f1-4)" 
    echo $name
    metaquast.py -o ../../../05_metaQUAST_NO_reference/"$host_name/""$name"_metaQuast contigs.fasta
    cd - > /dev/null  # return to previous dir
    pwd 
  done 
done

```
trifolium took longer than others, what kinds of samples were in this?
The output is 
`Floral_metagenomics/data/05_metaQUAST_NO_reference/"host_group"/"sample"/*`
These outputs only show number of contigs (`report.html` and legth stats
`basic_stats/`), and contig lengths 



### 6. MetaBat2

First, rename and group all of the fasta.contigs from the metaspades step into 
one folder, `06_sample_contigs`. This is so that there's one directory 
`06_sample_contigs` with all assembled contigs that contains the host 
(dandelion, apple, etc) in the file name, instead of branched into 
subdirectories by host name. 


#### 6.1 metabat prep
```{bash reorganize_sample_contigs}

cd /local/workdir/Daphne/Floral_metagenomics/data/

declare -a host_array=(
[0]=1_Chichorium    
[1]=2_Malus
[2]=3_Lotus
[3]=4_Ranunculus
[4]=5_Trifolium   
[5]=6_Matricaria
[6]=7_Alliaria
[7]=8_Solidago
[8]=9_Lonicera
[9]=10_Linaria
)

# rename and group all fasta.contigs from 04_metaspades into 06_sample_contigs
for host_name in ${host_array[@]}; do
  echo $host_name
  for sample in 04_metaspades"/$host_name/"*; do
    cd $sample
    name="$(basename "$sample" | cut -d'_' -f1-4)" 
    cp ./contigs.fasta ../../../06_sample_contigs/"$name"_"$host_name".fasta
    cd - > /dev/null  # return to previous dir
  done 
done
```
outputs are in `Floral_metagenomics/data/06_sample_contigs`



After the  metaspade fasta files have been renamed, we need to prepare the depth file, but it needs to be first reformatted with sam and bam.

.sam -> .bam -> depth.txt

#### 6.2 sam bam, depth files
```{bash sam_bam_depth, eval = FALSE}

```
















Then, we write metabat
NOTE: I may need a depth file, but I don't include it right now
```{bash metabat2, eval = FALSE}

for sample in 06_sample_contigs/*; do 
  singularity run --bind $PWD --pwd $PWD /programs/metabat-2.16/metabat.sif runMetaBat.sh -i $sample -a depth.txt -o 06_MetaBat2/
done

# -i = input contigs from metaspades (necessary)
# -a = depth file (necessary)
# -o = output folder, called 06_MetaBat2 (necessary)

```







